File: 0077_combinations/main.cpp
#include <vector>

class Solution {
private:
    void dfs(int n, int k, int startIdx, std::vector<int>& comb, std::vector<std::vector<int>>& res) {
        if(comb.size() == k) {
            res.push_back(comb);
            return;
        }

        for(int i = startIdx; i <= n; ++i) {
            comb.push_back(i);
            dfs(n, k, i + 1, comb, res);
            comb.pop_back();
        }
    }
public:
    std::vector<std::vector<int>> combine(int n, int k) {
        std::vector<std::vector<int>> res;
        std::vector<int> subset;
        dfs(n, k, 1, subset, res);
        return res;
    }
};

File: 0077_combinations/main.py
from typing import List

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        def dfs(startIdx: int, comb):
            if len(comb) == k:
                res.append(comb[:])
                return

            for i in range(startIdx, n + 1):
                comb.append(i)
                dfs(i+1, comb)
                comb.pop()
        dfs(1, [])
        return res
